piv/pcsc.go

```diff
@@ -48,16 +49,91 @@ func (e *scErr) Error() string {
 	return fmt.Sprintf("unknown pcsc return code 0x%08x", e.rc)
 }
 
-type apduErr struct {
+// VerifyErr is an error indicating an authentication error occurred (wrong PIN).
+type VerifyErr struct {
+	Retries int
+}
+
+func (v VerifyErr) Error() string {
+	if v.Retries > 0 {
+		return fmt.Sprintf("verification failed (%d retries remaining)", v.Retries)
+	}
+	return "verification failed"
+}
+
+// ErrNotFound is returned when the requested object on the smart card is not found.
+var ErrNotFound = errors.New("data object or application not found")
+
+// ErrBlocked is returned when the authentication method on the smart card is blocked,
+// such as when you exhaust PIN retries.
+var ErrBlocked = errors.New("authentication method blocked")
+
+// ApduErr is an error interacting with the PIV application on the smart card.
+// This error may wrap more accessible errors, like ErrNotFound, ErrBlocked, or
+// an instance of VerifyErr, so callers are encouraged to use errors.Is and errors.As
+// for these common cases.
+type ApduErr struct {
```

Comment:
Commented but I don't think this should be exported, though I'm fine with improving error messages for Error() and it supporting Unwrap().

Reply:
I returned this to its unexported form.  However, callers will not be able to access the PIV error codes without parsing it out of the string message.

Reply:
Thanks. I think that's good for now. We can change it in the future if we want :)